pkg/controller/node/cidr_allocator.go

```diff
@@ -17,144 +17,169 @@ limitations under the License.
+func (r *rangeAllocator) AllocateOrOccupyCIDR(obj interface{}) error {
+	node := obj.(*api.Node)
+	if node.Spec.PodCIDR != "" {
+		_, podCIDR, err := net.ParseCIDR(node.Spec.PodCIDR)
+		if err != nil {
+			return fmt.Errorf("failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
+		}
+		if err := r.cidrs.occupy(podCIDR); err != nil {
+			return fmt.Errorf("failed to mark cidr as occupied: %v", err)
+		}
+		return nil
+	}
+	podCIDR, err := r.cidrs.allocateNext()
 	if err != nil {
-		return err
+		recordNodeStatusChange(r.recorder, node, "CIDRNotAvailable")
+		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
 
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	r.used.SetBit(&r.used, used, 0)
-
+	glog.V(10).Infof("Putting node %s with CIDR %s into the work queue", node.Name, podCIDR)
+	r.nodeCIDRUpdateChannel <- nodeAndCIDR{
+		nodeName: node.Name,
+		cidr:     podCIDR,
+	}
 	return nil
 }
 
-func (r *rangeAllocator) MaxCIDRs() int {
-	return r.maxCIDRs
-}
-
-func (r *rangeAllocator) Occupy(cidr *net.IPNet) (err error) {
-	begin, end := 0, r.maxCIDRs
-	cidrMask := cidr.Mask
-	maskSize, _ := cidrMask.Size()
+// ReleaseCIDR released the CIDR of the removed node
+func (r *rangeAllocator) ReleaseCIDR(obj interface{}) error {
+	node := obj.(*api.Node)
 
-	if !r.clusterCIDR.Contains(cidr.IP.Mask(r.clusterCIDR.Mask)) && !cidr.Contains(r.clusterCIDR.IP.Mask(cidr.Mask)) {
-		return fmt.Errorf("cidr %v is out the range of cluster cidr %v", cidr, r.clusterCIDR)
+	if node.Spec.PodCIDR == "" {
+		return nil
 	}
 
-	if r.clusterMaskSize < maskSize {
-		subNetMask := net.CIDRMask(r.subNetMaskSize, 32)
-		begin, err = r.getIndexForCIDR(&net.IPNet{
-			IP:   cidr.IP.To4().Mask(subNetMask),
-			Mask: subNetMask,
-		})
-		if err != nil {
-			return err
-		}
-
-		ip := make([]byte, 4)
-		ipInt := binary.BigEndian.Uint32(cidr.IP) | (^binary.BigEndian.Uint32(cidr.Mask))
-		binary.BigEndian.PutUint32(ip, ipInt)
-		end, err = r.getIndexForCIDR(&net.IPNet{
-			IP:   net.IP(ip).To4().Mask(subNetMask),
-			Mask: subNetMask,
-		})
-		if err != nil {
-			return err
-		}
+	_, podCIDR, err := net.ParseCIDR(node.Spec.PodCIDR)
+	if err != nil {
+		return fmt.Errorf("Failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
 	}
 
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	for i := begin; i <= end; i++ {
-		r.used.SetBit(&r.used, i, 1)
+	glog.V(4).Infof("recycle node %s CIDR %s", node.Name, podCIDR)
+	if err = r.cidrs.release(podCIDR); err != nil {
+		return fmt.Errorf("Failed to release cidr: %v", err)
 	}
-
-	return nil
+	return err
 }
 
-func (r *rangeAllocator) getIndexForCIDR(cidr *net.IPNet) (int, error) {
-	cidrIndex := (binary.BigEndian.Uint32(r.clusterIP) ^ binary.BigEndian.Uint32(cidr.IP.To4())) >> uint32(32-r.subNetMaskSize)
+func (r *rangeAllocator) filterOutServiceRange(serviceCIDR *net.IPNet) {
+	if !r.clusterCIDR.Contains(serviceCIDR.IP.Mask(r.clusterCIDR.Mask)) && !serviceCIDR.Contains(r.clusterCIDR.IP.Mask(serviceCIDR.Mask)) {
```

Comment:
Can you please add comment what exactly you are trying to achieve here? [those operations are not trivial].